### PR TITLE
Fix Tim Join Logic™ for good

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

File: src/main/scala/is/hail/variant/MatrixTable.scala [hunk 1]
```diff
@@ -750,41 +750,13 @@ class MatrixTable(val hc: HailContext, val ast: MatrixIR) extends JoinAnnotator
...
-      val (t, ins) = if (expr != null) {
-        val ec = EvalContext(Map(
-          "sa" -> (0, colType),
-          "table" -> (1, joinSignature)))
-        Annotation.buildInserter(expr, colType, ec, Annotation.SAMPLE_HEAD)
-      } else insertSA(joinSignature, root)
-
-      (t, (a: Annotation, toIns: Annotation) => ins(a, f(toIns)))
-    }
+    val (finalType, inserter) = colType.structInsert(kt.valueSignature, List(root))
```
Comment (danking) [inline src/main/scala/is/hail/variant/MatrixTable.scala]: This all looks great except I'm confused about why the to-be-inserted-struct-type does not depend on the value of `product`. Can you explain that to me? The tests pass, so I must misunderstand something (or we need more exhaustive tests).

Comment (tpoterba) [reply src/main/scala/is/hail/variant/MatrixTable.scala]: this definitely looks like a bug.

Comment (danking) [reply src/main/scala/is/hail/variant/MatrixTable.scala]: let's get some tests that expose it as a part of this PR too!

Comment (tpoterba) [reply src/main/scala/is/hail/variant/MatrixTable.scala]: yup, will do that first! I'm not sure we can do it in Python yet, but definitely in Scala.

Comment (tpoterba) [reply src/main/scala/is/hail/variant/MatrixTable.scala]: or maybe we can with view_join_x...

**Pull request merged**